provider/bidengine/order.go

```diff
@@ -29,52 +31,55 @@ type order struct {
 	lc  lifecycle.Lifecycle
 }
 
-func newOrder(e *service, oid mtypes.OrderID, bid *mtypes.Bid) (*order, error) {
-
+func newOrder(svc *service, oid mtypes.OrderID, bid *mtypes.Bid, pricingStrategy BidPricingStrategy) (*order, error) {
 	// Create a subscription that will see all events that have not been read from e.sub.Events()
-	sub, err := e.sub.Clone()
+	sub, err := svc.sub.Clone()
 	if err != nil {
 		return nil, err
 	}
 
-	session := e.session.ForModule("bidengine-order")
+	session := svc.session.ForModule("bidengine-order")
 
 	log := session.Log().With("order", oid)
 
 	order := &order{
-		order:   oid,
-		bid:     bid,
-		session: session,
-		cluster: e.cluster,
-		bus:     e.bus,
-		sub:     sub,
-		log:     log,
-		lc:      lifecycle.New(),
+		orderID:         oid,
+		bid:             bid,
+		session:         session,
+		cluster:         svc.cluster,
+		bus:             svc.bus,
+		sub:             sub,
+		log:             log,
+		lc:              lifecycle.New(),
+		pricingStrategy: pricingStrategy,
 	}
 
 	// Shut down when parent begins shutting down
-	go order.lc.WatchChannel(e.lc.ShuttingDown())
+	go order.lc.WatchChannel(svc.lc.ShuttingDown())
 
 	// Run main loop in separate thread.
 	go order.run()
 
 	// Notify parent of completion (allows drain).
 	go func() {
 		<-order.lc.Done()
-		e.drainch <- order
+		svc.drainch <- order
 	}()
 
 	return order, nil
 }
 
 func (o *order) run() {
 	defer o.lc.ShutdownCompleted()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```

Comment:
you'll want to `cancel()` right after leaving the main event loop to start draining everything, not here.